examples/eks/eks_cluster_assumerole/castai.tf

```diff
@@ -1,46 +1,34 @@
+# 3. Connect EKS cluster to CAST AI.
+
 locals {
   role_name = "castai-eks-role"
 }
 
-provider "helm" {
-  kubernetes {
-    host                   = data.aws_eks_cluster.eks.endpoint
-    cluster_ca_certificate = base64decode(data.aws_eks_cluster.eks.certificate_authority[0].data)
-    token                  = data.aws_eks_cluster_auth.eks.token
-  }
-}
+# Configure Data sources and providers required for CAST AI connection.
+data "aws_caller_identity" "current" {}
 
-provider "aws" {
-  region = var.cluster_region
+data "castai_eks_user_arn" "castai_user_arn" {
+  cluster_id = castai_eks_clusterid.cluster_id.id
 }
 
 provider "castai" {
   api_token = var.castai_api_token
-  api_url   = var.castai_api_url
 }
 
-provider "kubernetes" {
-  host                   = data.aws_eks_cluster.eks.endpoint
-  cluster_ca_certificate = base64decode(data.aws_eks_cluster.eks.certificate_authority[0].data)
-  token                  = data.aws_eks_cluster_auth.eks.token
-}
-
-data "aws_caller_identity" "current" {}
-
-data "aws_eks_cluster_auth" "eks" {
-  name = module.eks.cluster_id
-}
-
-resource "castai_eks_clusterid" "cluster_id" {
-  account_id   = data.aws_caller_identity.current.account_id
-  region       = var.cluster_region
-  cluster_name = var.cluster_name
-}
-
-data "castai_eks_user_arn" "castai_user_arn" {
-  cluster_id = castai_eks_clusterid.cluster_id.id
+provider "helm" {
+  kubernetes {
+    host                   = module.eks.cluster_endpoint
+    cluster_ca_certificate = base64decode(module.eks.cluster_certificate_authority_data)
+      exec {
+        api_version = "client.authentication.k8s.io/v1beta1"
+        command     = "aws"
+        # This requires the awscli to be installed locally where Terraform is executed.
+        args = ["eks", "get-token", "--cluster-name", module.eks.cluster_name]
```

Comment:
Why we can't use `data.aws_eks_cluster_auth.eks.token`?

Reply:
`data.aws_eks_cluster_auth.eks.token` can only be used if EKS cluster exists. If `terraform apply` is run when cluster does not exist, Terraform fails with error that it cannot receive this data source as it doesn't exist. 
The same approach is used in AWS community module - https://github.com/terraform-aws-modules/terraform-aws-eks/blob/master/examples/complete/main.tf#L5